### PR TITLE
Improve component initialization for services

### LINKED ISSUE #362: Streamline component re-initialization
Minor parameter changes should not trigger component initialization.

### PR DESCRIPTION
**Pull request by kpfadnis**
# Pull Request      

## What does this PR do?

Closes #362     

Notes:
- Replace `(issue)` **above** ↑↑↑ with the issue this PR closes to automatically link the two.
This must be done when the PR is created.
- Add multiple `Closes #(issue)` as needed.
- If this PR is work towards but does not close an issue, simply tag the issue without mentioning `Closes`.

## Description

**Describe** the changes proposed by this PR below to give the reviewer context below ↓↓↓

1. Added hash functions to each baseline components like reader and retriever which is used in establishing need to create new instance.
2. Added additional metadata in field definition for each baseline components to further simplify service level consumption. 

Logic:
1. With every request for retriever or reader based operation, new instance of retriever or reader is created only when hash values are different. 

### Request Review

Be sure to **request a review** from one or more reviewers (unless the PR is to an unprotected branch).

## Versioning

When opening a PR to make changes to PrimeQA (i.e. `primeqa/`) master, be sure to increment the version following
[semantic versioning](https://semver.org/).  The VERSION is stored [here](https://github.com/primeqa/primeqa/blob/main/VERSION) 
and is incremented using `bump2version {patch,minor,major}` as described in the (development guide documentation)[development.html].

 - [ ] Have you updated the VERSION?
 - [ ] Or does this PR not change the `primeqa` package or was not into master?

After pulling in changes from master to an existing PR, ensure the VERSION is updated appropriately.
This may require bumping the version again if it has been previously bumped.

If you're not quite ready yet to post a PR for review, feel free to open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### After Merging
If merging into master and VERSION was updated, after this PR is merged:
- [ ] [Create a release](https://docs.github.com/en/github/administering-a-repository/releasing-projects-on-github/managing-releases-in-a-repository#creating-a-release) from the master with version equal to [VERSION](https://github.com/primeqa/primeqa/blob/main/VERSION)
- [ ] Not merging to master or VERSION not updated

### Checklist

Review the following and mark as completed:
- [x] [Tag an issue or issues](#what-does-this-pr-do) this PR addresses.
- [x] Added [description](#description) of changes proposed.
- [x] Review requested as [appropriate](#request-review).
- [ ] Version bumped as [appropriate](#Versioning).
- [x] New classes, methods, and functions documented.
- [ ] Documentation for modified code is updated.
- [ ] Built documentation to confirm it renders as expected (see [here](development.md)).
- [x] Code cleaned up and commented out code removed.
- [ ] Tests added to ensure all functionalities tested at >= 60% unit test coverage (see [here](development.md)).
- [ ] Code cleaned up and commented out code removed.
- [ ] Release created [as needed](#after-merging) after merging.

### CODE REVIEW

**Review by bhavani105 — COMMENTED** [review]
LGTM
Left some comments on additional  documentation and logging

File: primeqa/services/rest_server/server.py [hunk 1]
```diff
@@ -89,10 +89,26 @@ def run(self) -> None:
...
+                if mobj:
+                    error_code = mobj.group(1).strip()
+                    error_message = mobj.group(2).strip()
+                else:
+                    error_code = 500
+
+                raise HTTPException(
+                    status_code=500,
+                    detail={"code": error_code, "message": error_message},
+                ) from None
```
[the remark this answers is not in the document]
Comment (kpfadnis) [reply primeqa/services/rest_server/server.py:111]: Error message is passed on to caller so logging can be skipped.

**Review by bhavani105 — COMMENTED** [review]
LGTM

Can we remove the BM25Retriever stub and GenerativeReaders stub

**Review by bhavani105 — APPROVED** [review]
LGTM
Pls bumpversion before merge

**Review by franzmpub — APPROVED** [review]
LGTM

File: primeqa/pipelines/components/reader/extractive.py [hunk 2]
```diff
@@ -28,14 +29,20 @@ class ExtractiveReader(ReaderComponent):
...
+            a. api_support (bool, optional): If set to True, that parameter is exposed via service layer. Defaults to False.
+            b. exclude_from_hash (bool,optional): If set to True, that parameter is not considered while building the hash representation for the object. Defaults to False.
+
 
     Returns:
         _type_: _description_
     """
 
     model: str = field(
         default="PrimeQA/tydiqa-primary-task-xlm-roberta-large",
```
Comment (avisil) [inline primeqa/pipelines/components/reader/extractive.py:44]: can we have @rongzhangibm 's model uploaded and point to that one instead? That model is at least a couple of points better.

**Review by avisil — APPROVED** [review]
LGTM : approved but left some comments - we should change the uploaded default reader model.

**Pull request merged**